### PR TITLE
Fix use case t4rec

### PR DESCRIPTION
**Pull request by radekosmulski**
Remove incorrect line.

### CODE REVIEW

**Comment by rnyak** [comment]
@radekosmulski there is an error in this unit test tests/unit/tf/examples/test_usecase_transformers_next_item_prediction.py. we need to figure out why. something might have changed, I could repro error at my end with different dataset.

File: examples/usecases/transformers-next-item-prediction.ipynb [hunk 1]
```diff
@@ -538,14 +538,6 @@
     "We are now ready to train our model."
```
Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:538]: <pre class="ql-syntax" spellcheck="false">Line #11.    get_booking('/workspace/data')
</pre><p>This does data preprocessing under the hood as well, therefore, it is time and mostly memory consuming. is it possible that you only download the raw dataset since you are doing the preprocessing already in the following cells?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/943/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:538]: <p>We need to add an MLPBlock after the Input Block for projecting the input output dim to XLNET d_model dim. I recommend to modify the code as below:</p><p> </p><pre class="ql-syntax" spellcheck="false">dmodel=48

mlp_block = mm.MLPBlock(
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [128,dmodel],
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; activation='relu',
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; no_activation_last_layer=True,
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; )

model = mm.Model(
&nbsp; &nbsp; mm.InputBlockV2(
&nbsp; &nbsp; &nbsp; &nbsp; seq_schema,
&nbsp; &nbsp; &nbsp; &nbsp; embeddings=mm.Embeddings(
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; train_set_processed.schema.select_by_tag(Tags.CATEGORICAL), sequence_combiner=None
&nbsp; &nbsp; &nbsp; &nbsp; ),
&nbsp; &nbsp; ),
&nbsp; &nbsp; mlp_block,
&nbsp; &nbsp; mm.XLNetBlock(d_model=dmodel, n_head=4, n_layer=2,&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; pre=mm.SequentialBlock([mm.SequenceMaskLastInference(), mm.ReplaceMaskedEmbeddings()]),
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; post="inference_hidden_state",
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;),
&nbsp; &nbsp; mm.CategoricalOutput(
&nbsp; &nbsp; &nbsp; &nbsp; train_set_processed.schema.select_by_name(target),
&nbsp; &nbsp; &nbsp; &nbsp; default_loss="categorical_crossentropy",
&nbsp; &nbsp; ),
)
</pre><p>you can change the <code>dmodel</code> as you like. if you update the code like that then you can remove this line from the unit test <code>tb.cells[28].source = tb.cells[28].source.replace("d_model=64", "d_model=40")</code> . The reason it was giving error is bcs MLP block is missing and that creates dimension mismatch.</p><p> Please also note that I updated the pre arg as well.   it is to support inference.   the input sequence and its corresponding mask should be  extended by one position at the end to account for the next-item (target) position.     To do this, you should set SequenceMaskLastInference  as a pre-layer of    <code>ReplaceMaskedEmbeddings()</code> using the sequential-block. we dont have to add that in this example since this is only doing training and eval. you can remove it if you want to.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/943/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:538]: I am quite fond of relying on the functionality in our library to get the data, makes the notebooks much cleaner (fewer lines of code) and introduces users to the functionality of our library, but happy to change this if that is what we would like to do here

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:538]: thank you @rnyak, this is great! 🙂 made the changes

File: examples/usecases/transformers-next-item-prediction.ipynb [hunk 2]
```diff
@@ -74,18 +74,18 @@
      "text": [
```
Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:74]: <pre class="ql-syntax" spellcheck="false">Line #16.                      pre=mm.SequentialBlock([mm.SequenceMaskLastInference(), mm.ReplaceMaskedEmbeddings()]),
</pre><p>Radek do mind to add one liner explanation about SequenceMaskLastInference.  why we added it and what it does.. you can say simply write something like " SequenceMaskLastInference() -  is set to support inference. The input sequence and its corresponding mask should be extended by one position at the end to account for the next-item (target) position. To do this, we  set SequenceMaskLastInference as a pre-layer of&nbsp;<code>ReplaceMaskedEmbeddings()</code>&nbsp;using the sequential-block." </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/943/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:74]: hi @rnyak! Thank you for the comment! 🙂 I integrated this piece of information with what we already had there, please see below

Could you please verify if this is looking okay? Thank you!

image

Comment (rnyak) [reply examples/usecases/transformers-next-item-prediction.ipynb:77]: @radekosmulski it is not really selecting last item in the sequence. it is extending the original input sequence  and its corresponding mask  by one position at the end to account for the next-item (`target`) position at the `inference step` so not at the model.fit().  I recommend you to remove that arg since it is not required for model training and evaluation. Better to remove it :)

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:77]: that is a great point! apologies, didn't realize that this is how it works, remove it now

Comment (rnyak) [reply examples/usecases/transformers-next-item-prediction.ipynb:77]: <p>thanks! I am really sorry for back and forth. I think since you removed the&nbsp;<code>SequenceMaskLastInference</code>&nbsp;you can set the XLNETBlock like that, similar to what you were originally doing.</p><pre class="ql-syntax" spellcheck="false">mm.XLNetBlock(d_model=dmodel, n_head=4, n_layer=2, pre=mm.ReplaceMaskedEmbeddings())
</pre><p> </p>

**Comment by rnyak** [comment]
rerun tests

**Comment by rnyak** [comment]
rerun tests

**Review by rnyak — APPROVED** [review]
